docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,256 @@
+general any moving tag, as that renders the comparison useless. In the _SaaS_
+stage environment that variable is already using a tag, so that shouldn't be a
+problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The service will check during startup the value of the `AGENT_DOCKER_IMAGE` and
+it will write a warning to the log if it is using the `:latest` tag.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
+> process of being upgraded to a newer version, please wait, this could take
+> a few minutes.
+
+A new `upgrade-agent` step will be added to the service and to the agent.
+
+When the service receives a request for next steps from a host it will check if
+that host is in one of the following states:
+
+- `added-to-existing-cluster`
+- `binding`
+- `disabled-unbound`
+- `disabled`
+- `disconnected-unbound`
+- `disconnected`
+- `discovering-unbound`
+- `discovering`
+- `installed`
+- `installing-in-progress`
+- `insufficient-unbound`
+- `insufficient`
+- `known-unbound`
+- `known`
+- `pending-for-input`
+- `preparing-failed`
+- `preparing-successful`
+- `resetting-pending-user-action`
+- `resetting`
+- `unbinding-pending-user-action`
+- `unbinding`
+
+If the host is in one of the above states then the service will check the result
+of the `compatible-agent` validation. If the validation passed then the service
+will return the usal next steps. If the validaton didn't pass then the service
+will return only the `upgrade-agent` step, replacing all the usual steps.
+
+If on the other hand the host is in one of the followin states:
+
+- `preparing-for-installation`
+- `installing`,
+- `installing-in-progress`
+- `error`
+- `cancelled`
+
+Then the service will return the usual next steps regardless of the result of
+the validation. The reason for this is that we don't want to interrupt ongoing
+installations of log gathering.
+
+The agent will be changed so that it sends the full image reference in the
+`discovery_agent_version` header.
+
+When the agent receives the `agent-upgrade` step it will do nothing and return
+an exit code indicating that the agent needs to be upgraded. When receiving that
+exit code the next step runner will also exit, with the same code. The main
+program will react to that exit code starting the next step runner with the new
+image.
+
+The `upgrade-agent` step is also (similar to register) something we must not
+break in order to allow agent upgrade.
+
+### Risks and Mitigations
```

Comment:
@avishayt is your suggestion to have another mini state machine for upgrades, parallel to the regular host state machine? Or just a free text field just for display purposes?

Reply:
a free text field just for display purposes

Reply:
@avishayt there will be events generated by the server when the upgrade starts, when it finishes, and when it fails, something like this:

![image](https://user-images.githubusercontent.com/1423971/176187313-8086d6b9-9a4f-4b92-a859-517752caa38c.png)

So there will already be information available for debugging. That will not be included in the validation message, only in the events. Do you think it is necessary to also have that information in the validation message?

Reply:
Let's leave this as an open question and start without it.  The problem is that when using the k8s API events aren't really accessible to the user.  It would be best to either make the validation stand-alone or start sending k8s events.  Let's discuss with a wider forum offline.

Reply:
Added it to the _Open Questions_ section of the document.